crate_universe/extension.bzl

```
@@ -0,0 +1,224 @@
+"""Module extension for generating third-party crates for use in bazel."""
+
+load("@bazel_tools//tools/build_defs/repo:git.bzl", "new_git_repository")
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+load("//crate_universe:defs.bzl", _crate_universe_crate = "crate")
+load("//crate_universe/private:crates_vendor.bzl", "CRATES_VENDOR_ATTRS", "generate_config_file", "generate_splicing_manifest")
+load("//crate_universe/private:generate_utils.bzl", "render_config")
+load("//crate_universe/private/module_extensions:cargo_bazel_bootstrap.bzl", "get_cargo_bazel_runner")
+
+def _generate_repo_impl(repo_ctx):
+    for path, contents in repo_ctx.attr.contents.items():
+        repo_ctx.file(path, contents)
+
+_generate_repo = repository_rule(
+    implementation = _generate_repo_impl,
+    attrs = dict(
+        contents = attr.string_dict(mandatory = True),
+    ),
+)
+
+def _generate_annotations(module_ctx, annotation_files):
+    """Given a list of annotation files, merge them into a single json object.
+
+    For an example json file, see examples/bzlmod/hello_world/annotations.json.
+    """
+    annotations = {}
+
+    for file in annotation_files:
+        content = json.decode(module_ctx.read(file))
+        for name, annotations_for_crate in content.items():
+            for annotation in annotations_for_crate:
+                annotation = _crate_universe_crate.annotation(**annotation)
+                annotations.setdefault(name, [])
+                annotations[name].append(annotation)
+
+    return annotations
+
+def _generate_hub_and_spokes(module_ctx, cargo_bazel, mod_path, cfg):
+    annotations = _generate_annotations(module_ctx, cfg.annotation_files)
+    tag_path = mod_path
+    repo_name = "crates"
+    if cfg.suffix:
+        tag_path = mod_path.get_child(cfg.suffix)
+        repo_name += "_" + cfg.suffix
+
+    cargo_lockfile = module_ctx.path(cfg.cargo_lockfile)
+
+    rendering_config = json.decode(render_config(
+        regen_command = "Run 'cargo update [--workspace]'",
+    ))
+    config_file = tag_path.get_child("config.json")
+    module_ctx.file(
+        config_file,
+        executable = False,
+        content = generate_config_file(
+            module_ctx,
+            mode = "remote",
+            annotations = annotations,
+            generate_build_scripts = cfg.generate_build_scripts,
+            supported_platform_triples = cfg.supported_platform_triples,
+            generate_target_compatible_with = True,
+            repository_name = repo_name,
+            output_pkg = repo_name,
+            workspace_name = repo_name,
+            generate_binaries = cfg.generate_binaries,
+            rendering_config = rendering_config,
+        ),
+    )
+
+    manifests = {module_ctx.path(m): m for m in cfg.manifests}
+    splicing_manifest = tag_path.get_child("splicing_manifest.json")
+    module_ctx.file(
+        splicing_manifest,
+        executable = True,
```

This is a json file, right?

```suggestion
        executable = False,
```